// hazelcast-client/lib/codec/ImmutableLazyDataList.js
"use strict";
/*
 * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var ImmutableLazyDataList = /** @class */ (function () {
    function ImmutableLazyDataList(response, toObjectFunction) {
        this.response = response;
        this.toObjectFunction = toObjectFunction;
    }
    ImmutableLazyDataList.prototype.contains = function (item) {
        return this.response.contains(item);
    };
    ImmutableLazyDataList.prototype.size = function () {
        return this.response.length;
    };
    ImmutableLazyDataList.prototype.get = function (index) {
        var element = this.response[index];
        return [element.key, element.value];
    };
    ImmutableLazyDataList.prototype.populate = function () {
        //TODO
    };
    ImmutableLazyDataList.prototype.equal = function () {
        //TODO
        return false;
    };
    return ImmutableLazyDataList;
}());
module.exports = ImmutableLazyDataList;
